src/abstractdataframe/reshape.jl

```diff
@@ -185,39 +193,71 @@ long = stack(wide)
-
 """
-function unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)
-    # `rowkey` integer indicating which column to place along rows
-    # `colkey` integer indicating which column to place along column headers
-    # `value` integer indicating which column has values
-    refkeycol = CategoricalArray{Union{eltype(df[rowkey]), Missing}}(df[rowkey])
-    # make sure we report only levels actually present in rowkey column
-    # this is consistent with how the other unstack method based on groupby works
+function unstack(df::AbstractDataFrame, rowkeys::Int, colkey::Int, value::Int)
+    refkeycol = copy(CategoricalArray{Union{eltype(df[rowkeys]), Missing}}(df[rowkeys]))
     droplevels!(refkeycol)
-    valuecol = df[value]
     keycol = CategoricalArray{Union{eltype(df[colkey]), Missing}}(df[colkey])
+    valuecol = df[value]
     Nrow = length(refkeycol.pool)
     Ncol = length(keycol.pool)
-    payload = DataFrame(Any[similar_missing(valuecol, Nrow) for i in 1:Ncol], map(Symbol, levels(keycol)))
-    nowarning = true
+    df2 = DataFrame(Any[similar_missing(valuecol, Nrow) for i in 1:Ncol], map(Symbol, levels(keycol)))
+    _unstack(df, rowkeys, colkey, value, keycol, valuecol, df2, refkeycol)
+end
+
+function _unstack(df::AbstractDataFrame, rowkeys::Int,
+                  colkey::Int, value::Int, keycol, valuecol, df2, refkeycol)
+    Nrow = length(refkeycol.pool)
+    Ncol = length(keycol.pool)
+    hadmissing = false # have we encounered missing in refkeycol
+    mask_filled = falses(Ncol, Nrow) # has a given [col,row] entry been filled?
+    nowarning = true # do we print duplicate entries warning
+    nowarning2 = true # do we print missing in keycol
+    keycol_pool = Vector{Int}(CategoricalArrays.order(keycol.pool))
+    refkeycol_pool = Vector{Int}(CategoricalArrays.order(refkeycol.pool))
     for k in 1:nrow(df)
-        j = Int(CategoricalArrays.order(keycol.pool)[keycol.refs[k]])
-        i = Int(CategoricalArrays.order(refkeycol.pool)[refkeycol.refs[k]])
-        if i > 0 && j > 0
-            if nowarning && !ismissing(payload[j][i])
-                warn("Duplicate entries in unstack.")
-                nowarning = false
+        keycol_refs = keycol.refs[k]
+        if keycol_refs == 0 # we have found missing in colkey
+            if nowarning2
+              warn("Missing value in colkey variable at row $k. Skipping.")
+              nowarning2 = false
+            end
+            continue # skip processing it
+        end
+        j = keycol_pool[keycol_refs]
+        refkeycol_refs = refkeycol.refs[k]
+        if refkeycol_refs == 0 # we have found missing in rowkeys
+            if !hadmissing # if it is the first time we have to add a new row
+                hadmissing = true
+                # we use the fact that missing is greater than anything
+                push!(df2, missings(ncol(df2)))
+                mask_filled = hcat(mask_filled, falses(ncol(df2)))
```

Comment:
Better allocate the column for missing values unconditionally, even if it's not used. Also doesn't the data frame column need a name?

Reply:
fixed `mask_filled`. As for `push!` - no it adds a row not a column (now it is changed anyway as I do create a `DataFrame` later in the code).